server/sonar-db-dao/src/main/java/org/sonar/db/issue/IssueDto.java

```diff
@@ -0,0 +1,726 @@
+      .setLocations((DbIssues.Locations) issue.getLocations())
+      .setMessage(issue.message())
+      .setGap(issue.gap())
+      .setEffort(issue.effortInMinutes())
+      .setResolution(issue.resolution())
+      .setStatus(issue.status())
+      .setSeverity(issue.severity())
+      .setManualSeverity(issue.manualSeverity())
+      .setChecksum(issue.checksum())
+      .setAssignee(issue.assignee())
+      .setRuleId(ruleId)
+      .setRuleKey(issue.ruleKey().repository(), issue.ruleKey().rule())
+      .setTags(issue.tags())
+      .setComponentUuid(issue.componentUuid())
+      .setComponentKey(issue.componentKey())
+      .setModuleUuid(issue.moduleUuid())
+      .setModuleUuidPath(issue.moduleUuidPath())
+      .setProjectUuid(issue.projectUuid())
+      .setProjectKey(issue.projectKey())
+      .setIssueAttributes(KeyValueFormat.format(issue.attributes()))
+      .setAuthorLogin(issue.authorLogin())
+      .setIssueCreationDate(issue.creationDate())
+      .setIssueCloseDate(issue.closeDate())
+      .setIssueUpdateDate(issue.updateDate())
+      .setSelectedAt(issue.selectedAt())
+
+      // technical dates
+      .setCreatedAt(now)
+      .setUpdatedAt(now);
+  }
+
+  /**
+   * On server side, we need component keys and uuid
+   */
+  public static IssueDto toDtoForServerInsert(DefaultIssue issue, ComponentDto component, ComponentDto project, int ruleId, long now) {
+    return toDtoForComputationInsert(issue, ruleId, now)
+      .setComponent(component)
+      .setProject(project);
+  }
+
+  public static IssueDto toDtoForUpdate(DefaultIssue issue, long now) {
+    // Invariant fields, like key and rule, can't be updated
+    return new IssueDto()
+      .setKee(issue.key())
+      .setType(issue.type())
+      .setLine(issue.line())
+      .setLocations((DbIssues.Locations) issue.getLocations())
+      .setMessage(issue.message())
+      .setGap(issue.gap())
+      .setEffort(issue.effortInMinutes())
+      .setResolution(issue.resolution())
+      .setStatus(issue.status())
+      .setSeverity(issue.severity())
+      .setChecksum(issue.checksum())
+      .setManualSeverity(issue.manualSeverity())
+      .setAssignee(issue.assignee())
+      .setIssueAttributes(KeyValueFormat.format(issue.attributes()))
+      .setAuthorLogin(issue.authorLogin())
+      .setRuleKey(issue.ruleKey().repository(), issue.ruleKey().rule())
+      .setTags(issue.tags())
+      .setComponentUuid(issue.componentUuid())
+      .setComponentKey(issue.componentKey())
+      .setModuleUuid(issue.moduleUuid())
+      .setModuleUuidPath(issue.moduleUuidPath())
+      .setProjectUuid(issue.projectUuid())
+      .setProjectKey(issue.projectKey())
+      .setIssueCreationDate(issue.creationDate())
+      .setIssueCloseDate(issue.closeDate())
+      .setIssueUpdateDate(issue.updateDate())
+      .setSelectedAt(issue.selectedAt())
+
+      // technical date
+      .setUpdatedAt(now);
+  }
+
+  public static IssueDto createFor(Project project, RuleDto rule) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Remove this use of "Project"; it is deprecated. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)